x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx

```diff
@@ -0,0 +1,269 @@
+      setHoverState({});
+    },
+    []
+  );
+
+  // When new, memoized handlers are created with useCallback,
+  // these references allow the old handlers to be removed
+  const updateHoverStateOnMousemoveRef = useRef<
+    ((event: mapboxgl.MapMouseEvent & mapboxgl.EventData) => void) | null
+  >(null);
+
+  const updateHoverStateOnMouseoutRef = useRef<
+    ((event: mapboxgl.MapMouseEvent & mapboxgl.EventData) => void) | null
+  >(null);
+
+  // initialization side effect, only runs once
+  useEffect(() => {
+    if (containerRef.current !== null) {
+      // set up Map object
+      const mapboxMap = new Map({
+        attributionControl: false,
+        container: containerRef.current,
+        style:
+          'https://tiles.maps.elastic.co/styles/osm-bright-desaturated/style.json',
+        ...initialMapboxOptionsRef.current
+      });
+      mapboxMap.dragRotate.disable();
+      mapboxMap.touchZoomRotate.disableRotation();
+      mapboxMap.addControl(
+        new NavigationControl({ showCompass: false }),
+        'top-left'
+      );
+
+      // set up Popup object
+      popupRef.current = new Popup({
+        closeButton: false,
+        closeOnClick: false
+      });
+
+      // only scroll zoom when key is pressed
+      const canvasElement = mapboxMap.getCanvas();
+      canvasElement.addEventListener('wheel', controlScrollZoomOnWheel);
+
+      mapboxMap.on('load', () => {
+        mapboxMap.addSource(CHOROPLETH_POLYGONS_SOURCE_ID, {
+          type: 'geojson',
+          data: initialGeojsonSourceRef.current
+        });
+        setMap(mapboxMap);
+      });
+
+      // cleanup function called when component unmounts
+      return () => {
+        canvasElement.removeEventListener('wheel', controlScrollZoomOnWheel);
+      };
+    }
+  }, [controlScrollZoomOnWheel]);
+
+  // side effect swaps the old handlers with new ones when they update
+  useEffect(() => {
+    if (map) {
+      if (
+        updateHoverStateOnMousemoveRef.current &&
+        updateHoverStateOnMouseoutRef.current
+      ) {
+        map.off('mousemove', updateHoverStateOnMousemoveRef.current);
+        map.off('mouseout', updateHoverStateOnMouseoutRef.current);
+      }
+      map.on('mousemove', updateHoverStateOnMousemove);
+      map.on('mouseout', updateHoverStateOnMouseout);
+      updateHoverStateOnMousemoveRef.current = updateHoverStateOnMousemove;
+      updateHoverStateOnMouseoutRef.current = updateHoverStateOnMouseout;
+    }
+  }, [map, updateHoverStateOnMousemove, updateHoverStateOnMouseout]);
+
+  // side effect replaces choropleth layer with new one on data changes
+  useEffect(() => {
+    if (map) {
+      // find first symbol layer to place new layer in correct order
+      const symbolLayer = (map.getStyle().layers || []).find(
+        ({ type }) => type === 'symbol'
+      );
+      if (map.getLayer(CHOROPLETH_LAYER_ID)) {
+        map.removeLayer(CHOROPLETH_LAYER_ID);
+      }
+      if (data.length) {
```

Comment:
Again, return early. As a reader when I encounter this I have to prepare myself for to branches (I always scroll to the bottom to get an overview early). If you return early I immediately know there is only a single case left which makes it easier to reason about up front:

```js
if (data.length === 0) {
  return;
}
```